### PR TITLE
bus/nes: Moved gc6in1 games over to 8237 board and removed old code.

### PR DESCRIPTION
**Pull request by 0kmg**

### CODE REVIEW

**Comment by 0kmg** [comment]
This is the (likely) last follow-up to a recent rewrite (#8481). Previously we had three implementations of so-called mapper 215, with most of the games not working at all on any of the implementations. The (working) code being removed here is mostly similar to the rewrite except it handles a subset of cases and reimplements parts of MMC3 by hard coding the scrambled addresses of these boards in an explicit switch. The rewrite simple does a table lookup and passes to the underlying MMC3 parent class.

**Review by cuavas — COMMENTED** [review]
Is there a reason for removing the PCB type from these items?  Do they not actually use this PCB?

**Comment by 0kmg** [comment]
tl;dr: the "pcb" feature isn't used and there's no evidence to show these labels are correct anyway.

AFAICT the "pcb" field is made up for nes.xml and needs to be checked over entirely. Perhaps @etabeta78 could confirm but my surmise from having worked on NES stuff recently and having peeked into the old SVN is that "pcb" is legacy from before things were slot-ified. They do not actually represent labels on the boards but represent cart-type (mapper essentially) and were used to connect SL items to the cart-type lookup. That has been superseded by the "slot" feature. For what it's worth there are few entries in nes.xml that have "pcb_model" and those probably are (at least close) to actual labels.

As for bootlegs like these, I have near zero info on PCBs or even cartridges. 8237 and 8237A seem to be labels that appeared on at least some PCB for some games identified with mapper 215. The only other clue I have is that NewRisingSun, the apparent ringleader on the NesDev Wiki, provides the label "823X" in his code, suggesting he's perhaps seen some similar boards. For these particular games it's also worth noting that maybe only the 6-in-1 actually exists on cart and the other three may be extracts from it.

Unfortunately, I have had no contact with anybody who can shed light on minor details like these since the NesDev forum has coincidentally been down since I started working on this stuff.

**Comment by cuavas** [comment]
The `pcb` feature is supposed to document the actual PCB type used in the cartridges, while the `slot` feature tells MAME which slot option to instantiate.  The former is a piece of documentation, the latter is a bit of a leaky abstraction, but necessary for how the NES emulation works at the moment.  There isn’t a one-to-one mapping between cartridge PCB types and MAME slot options (for an extreme example, SNES has [a large number of distinct cartridge PCBs](https://snescentral.com/pcblisting.php), but a relatively small number of distinct cartridge models from an emulation perspective).

The cartridge types for these items have been there since 2012, before MAME and MESS were merged.  I don’t know how well they were sourced.  If you don’t think they’re reliable, I’ll trust your judgement in removing them.

**Comment by 0kmg** [comment]
Yeah, that's basically what I sussed out over time about the "pcb" feature. Again though, the documentation is often wrong due to the fact that MESS appears to have loaded the cart type based on these strings previously. Random example from btoadsu:

`<feature name="slot" value="axrom" />`
`<feature name="pcb" value="NES-AOROM" />`
`<feature name="pcb_model" value="NES-AOROM-03" />`

A lot of stuff in nes.xml looks like that. I'm sure we can find a picture on bootgod's DB of the actual PCB with actual label with suffix "-03". I presume eventually we'd want to remove the redundancy and have key-value pairs "pcb" + correct label, since other software lists treat "pcb" as the actual type.

Anyway, how this relates back to the bootleg PCB stuff is that basically I usually don't know what if anything describes the "pcb" so I've been erring on the side of not including a probably incorrect guess.

**Comment by 0kmg** [comment]
As for how things were sourced ten or so years ago, it looks to me as if MESS simply ported over implementations from Nestopia/FCEUmm and assigned things cartridge types (three in this case) based on those names—there's no more authoritative knowledge there than "some emulators did it this way". This almost certainly represents a LACK of understanding of/access to the actual PCBs, separate guesswork of how to emulate separate games, and then never noticing the similarity between them.

That being said, I'm working from NesDev wiki details here: https://wiki.nesdev.com/w/index.php?title=INES_Mapper_215 and can only be certain as they are. As always, without PCB photos, schematics, etc I'd recommend caution that the information may represent some idealized hardware superset that doesn't exactly reflect reality. But NesDev on the whole is about documenting real hardware differences, so this is better than trusting old emulator code written only with "getting games to work" in mind.

**Comment by etabeta78** [comment]
A few info so to avoid confusion:

* about overall quality of the list
believe it or not, in 2013 I manually tested all the dumps when creating
the list. However, except for bootgod's incredible documentation effort,
info about bootleg carts were almost nonexistent.
I had to compose a puzzle by using: iNES mapper number (not even iNES2.0),
UNIF pcbs (sometimes wrongly documented), NEStopia and FCEUmm sources.
Sometimes they were contradictory, or dumps were given non-existing mapper
numbers and I had to go through trial and error in matching them to a PCB
type. The overall reason behind wrong PCBs and so on is just a lack of
information.
There were similar problems with a lot of Japanese Namco and Bandai
mappers, until a guy from nesdev boards took time for probing a few PCBs
and I could include their findings in MAME/MESS.

* about romanization
as before, not enough information was offered in 2013. I went through huge
lists of advertisements from Waixing and Nainjing and others in order to
get a decent mapping between Romanized titles, cart IDs and Chinese titles,
but I know neither Japanese (except for some Katakana) nor Chinese. There
had been no feedback so far, and being NES the list with the largest number
of Chinese bootleg games, it is definitely the greatest offender among
software lists

* about bootleg PCBs:
The main problem is that back in 2013, nobody was documenting properly the
bootleg carts at the PCB level, and rather they were assigning random new
iNES/UNIF mappers.
As such, when a cart was not documented by bootgod, I worked with what was
available: sometimes it was bit of info from Cah4e3, sometimes it was a
bunch of code from Nestopia, sometimes old spare txt with outdated info. I
love the research that has been performed in the past 8 years, but I had no
access to it when the list was created.
All I had available were tons of .nes dumps with unreliable mapper numbers,
released on chinese forums with no real reference about the hardware. I
don't doubt that a few were not assigned the correct boards, but the
alternative was not to document the dump at all.

* about pcb_types:
PCB and PCB_MODEL come from bootgod's database and have never been used to
load things. they have never been used to load things, but they document
more precisely what is on the cart. None of the chinese bootlegs had been
documented to such a level of detail when the list was created, so that I
added some 'pcb' line as a placeholder until proper documentation could
have become available (but for most of them, in 2018 there had been no
picture surfacing).
What we use to load, and it has been always used, is SLOT (which allows to
reduce the redundancy across different PCBs, among other things).


I really appreciate the job 0kmg is doing in rectifying the mistakes I made
due to lack of any info from dumpers. Feel free to ask me about
implementations detail, if you cannot follow any of the legacy stuff.


'etabeta'

**Comment by 0kmg** [comment]
Thank you, etabeta, for the additional information. That's pretty close to the overall picture I had built in my head based on combing over the source code, the software list, combing through nesdev and bootleg game forums, investigating dump origins like CaH4e3, occasional glances at Nestopia and things to figure out code origins, etc. (Though I am surprised about the SLOT thing because I didn't see SLOT in the old mess-svn—not that going over historical trivia matters.)

For what it's worth I don't really think of any improvements I've been able to add as rectifying mistakes. Given the nature of the NES emulation scene over time, the incredible breadth of unofficial software available for so many still undocumented famiclones, the until-recent lack of interest in documenting unofficial carts, and so on...it's pretty impressive what you have been able to catalog. For my part, I'm only able to contribute due to that existing base and due to the flurry of activity and documentation around rarer carts in say the last 4-5 years. Even still, as you say, the lack of information remains the biggest obstacle.
